data/src/main/java/com/eleks/data/firebase/source/impl/FirebaseDataSourceImpl.kt

```diff
@@ -1,9 +1,140 @@
 package com.eleks.data.firebase.source.impl
 
 import com.eleks.data.firebase.source.FirebaseDataSource
+import com.eleks.data.model.GroupDataModel
+import com.eleks.data.model.QuoteDataModel
+import com.eleks.data.model.ResultDataModel
+import com.eleks.data.model.SelectedGroupDataModel
+import com.google.firebase.auth.FirebaseAuth
+import com.google.firebase.firestore.FirebaseFirestore
+import com.google.firebase.firestore.ktx.toObject
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.flow.*
+import kotlinx.coroutines.suspendCancellableCoroutine
+import java.util.*
 import javax.inject.Inject
 import javax.inject.Singleton
 
 @Singleton
-class FirebaseDataSourceImpl @Inject constructor(): FirebaseDataSource {
-}
\ No newline at end of file
+class FirebaseDataSourceImpl @Inject constructor(
+    private val dbInstance: FirebaseFirestore,
+    private val authInstance: FirebaseAuth
+) : FirebaseDataSource {
+
+    private val _groupsFlow = MutableStateFlow<ResultDataModel<List<GroupDataModel?>>>(
+        ResultDataModel.success(null)
+    )
+
+    private val _userGroupsFlow = MutableStateFlow<ResultDataModel<List<GroupDataModel?>>>(
+        ResultDataModel.success(null)
+    )
+
+    private val _selectedGroupsFlow = MutableStateFlow<ResultDataModel<List<SelectedGroupDataModel?>>>(
+        ResultDataModel.success(null)
+    )
+
+    override val groupsFlow = _groupsFlow
+
+    override val userGroupsFlow = _userGroupsFlow
+
+    override val selectedGroupsFlow = _selectedGroupsFlow
+
+    override fun subscribeGroups() {
+        dbInstance.collection("groups")
+            .addSnapshotListener { value, error ->
+                if (error != null) {
+                    _groupsFlow.value = ResultDataModel.error(error.message ?: "")
+                    return@addSnapshotListener
+                }
+                if (value != null && !value.isEmpty) {
+                    val groups = mutableListOf<GroupDataModel>()
+                    for (doc in value) {
+                        groups.add(doc.toObject())
+                    }
+                    _groupsFlow.value = ResultDataModel.success(groups)
+                }
+            }
+    }
+
+    override fun subscribeUserGroups() {
+        if (authInstance.currentUser != null) {
+            dbInstance.collection("group_${authInstance.currentUser?.uid}")
```

Comment:
Sorry, I'm not quite sure I understood how it should be implemented? So, if a new group is created for user A, does it mean that new collection is also created(group_A)? This collection will contain only items specific for that user?
Don't you think it will bloat our collections too much? We should discuss it together with @maryan91